dora/src/baseline/cannon/codegen.rs

```diff
@@ -587,6 +597,55 @@ where
             .store_mem(bytecode_type.mode(), Mem::Local(offset), REG_RESULT.into());
     }
 
+    fn emit_jump_if(
+        &mut self,
+        bytecode: &BytecodeFunction,
+        src: Register,
+        bytecode_idx: &BytecodeIdx,
+        op: CondCode,
+    ) {
+        assert_eq!(bytecode.register(src), BytecodeType::Bool);
+
+        let bytecode_type = bytecode.register(src);
+        let offset = bytecode.offset(src);
+        self.asm
+            .load_mem(bytecode_type.mode(), REG_RESULT.into(), Mem::Local(offset));
+
+        let lbl = self.asm.create_label();
+        self.asm.test_and_jump_if(op, REG_RESULT, lbl);
+
+        let BytecodeIdx(bytecode_idx) = bytecode_idx;
+        match self.bytecode_to_address.get(&bytecode_idx) {
+            Some(&address) => {
+                self.asm.bind_label_to(lbl, address);
+            }
+            None => {
+                self.forward_jumps.push(ForwardJump {
+                    label: lbl,
+                    bytecode_idx: *bytecode_idx,
+                });
+            }
+        }
+    }
+
+    fn emit_jump(&mut self, _bytecode: &BytecodeFunction, bytecode_idx: &BytecodeIdx) {
+        let lbl = self.asm.create_label();
+        self.asm.jump(lbl);
+
+        let BytecodeIdx(bytecode_idx) = bytecode_idx;
+        match self.bytecode_to_address.get(&bytecode_idx) {
```

Comment:
Right now, `bytecode_idx` is always increasing by 1. In the future this will be the offset in the bytecode stream. That's why I think it would be safer to shield against bugs with wrong offsets: `if target < current_position { self.bytecode_to_address.get(&offset).expect("jump with wrong offset") } else { /* forward jump*/ }`

Reply:
Would be awesome if these two functions could share this code as well.

Reply:
For now, I only moved the code to a separate function, but didn't add the check. I will revisit this tomorrow.